### ISSUE TITLE
[inductor] Numerical inconsistency: polygamma + cumulative_trapezoid under torch.compile(dynamic=True) produces wrong results after Conv2d + GroupNorm

### ISSUE THREAD

**Issue by lijunqiang123** [pytorch/pytorch#185587]
### 🐛 Describe the bug

Under `torch.compile(dynamic=True)` , a model combining Conv2d → GroupNorm → polygamma → cumulative_trapezoid produces numerically incorrect outputs compared to eager mode, despite both completing without errors. The maximum absolute difference is `1420.2` , indicating a correctness bug in Inductor's lowering or fusion for this operator combination.

Eager and compiled outputs have the same shape but contain drastically different values — the compiler silently produces wrong results without any error or warning.

### Minimal repro
```
import torch
import torch.nn as nn

torch.set_grad_enabled(False)

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 16, kernel_size=3, padding=1)
        self.gn1 = nn.GroupNorm(num_groups=4, num_channels=16)

    def forward(self, x):
        x = self.conv1(x)
        x = self.gn1(x)
        x = torch.polygamma(0, x + 0.1)
        x = x.unsqueeze(2)
        x = torch.cumulative_trapezoid(x)
        return x

model = Model()
x = torch.randn(1, 3, 32, 32)

res_eager = model(*[x])
res_compiled = torch.compile(model, dynamic=True)(x)

print("Max diff:", (res_eager - res_compiled).abs().max().item())
# Output: Max diff: 1420.1875
# The max diff is non-deterministic (varies between 31 and 1420 across runs due to random input), but always present.
```
### Result
Both outputs have identical shape but the values differ by up to 1420 . The discrepancy is widespread across the tensor, not isolated to a few elements.

### Root cause hypothesis
One or both of these Inductor lowerings likely have a numerical correctness issue under symbolic shapes:

- `torch.polygamma(0, x)` — the digamma function, polygamma of order 0
- `torch.cumulative_trapezoid(x)` — cumulative trapezoidal integration
When `dynamic=True` , Inductor cannot statically determine the integration axis length, which may cause:

1. An incorrect decomposition or fusion of `polygamma` with surrounding ops
2. A numerically unstable approximation in `cumulative_trapezoid` under symbolic dimensions

### Additional context
- Both eager and compiled modes complete without any error — this is a silent correctness bug (INCON type).
- Removing dynamic=True or any single operator from the chain may mask the problem.
- The max diff is non-deterministic (varies between 31 and 1420 across runs due to random input), but always present.

### Versions

```
PyTorch version: 2.13.0.dev20260521+cu126
OS: Linux (Ubuntu 20.04)
Python: 3.12
CUDA: 12.6
```

cc @mruberry @kshitij12345 @chauhang @penguinwu @ezyang @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by jansel** [comment]
My agent says this appears to be intended torch.compile floating-point behavior for an ill-conditioned program rather than an Inductor lowering bug. I reproduced the mismatch, but isolating the operators showed polygamma and cumulative_trapezoid match eager for the same input. The first large divergence happens because GroupNorm differs by normal fp32 roundoff near x + 0.1 ~= 3.7e-5, extremely close to digamma's pole at zero; that tiny ~2e-8 input difference is amplified into a large polygamma output difference. TORCHINDUCTOR_EMULATE_PRECISION_CASTS=1 and dynamic=False did not remove the mismatch, so this does not look dynamic-shape-specific.

**Comment by lijunqiang123** [comment]
Thanks for your prompt and thorough analysis, and especially for isolating the components. That really helped me understand the root cause.

To follow up on this, I have a couple of questions:

1. Are there any recommended practices or more numerically stable alternatives for computing `polygamma` (or `digamma`) when inputs can land near the singular point (0)? For example, would adding a small epsilon to the input help, or is there a better mathematical reformulation?

2. Could this be mentioned in the PyTorch documentation for `polygamma` or `GroupNorm` as a caveat? A brief note about the stability boundary near singularities would be very helpful for others who might encounter similar issues in the future.

Again, thank you for your insight. I really appreciate the time you took to dig into this.

**Comment by jansel** [comment]
For 1, I'm not sure, but perhaps @claude could you answer that question.
